stage1: support per-app memory and cpu isolators

iaguis opened the pull request:
To allow systemd inside stage1 to set resource limits for apps we mount
the cgroup controllers hierarchy in rkt. We need this because
systemd-nspawn mounts cgroup controllers as read-only.

We mimic systemd-nspawn cgroup mounts with the difference that we
bind-mount the apps' subdirectories over themselves so they stay
read-write. We leave the mounting of systemd's cgroup hierarchy to
nspawn.

Addresses (partially) #811

jonboulle commented on stage1/init/init.go:
```diff
@@ -295,6 +307,165 @@ func forwardedPorts(pod *Pod) ([]networking.ForwardedPort, error) {
+		return fmt.Errorf("error mounting %q: %v", cgroupTmpfs, err)
+	}
+
+	// 4. Mount controllers
```
missing a 3

jonboulle commented on stage1/init/init.go:
```diff
@@ -295,6 +307,165 @@ func forwardedPorts(pod *Pod) ([]networking.ForwardedPort, error) {
+			if err != nil {
+				return fmt.Errorf("error getting unit name: %v", err)
+			}
+			subcgroup = filepath.Join(cPath, "system.slice", unit, "system.slice")
```
isn't this counter to https://github.com/coreos/rkt/issues/911#issuecomment-103853978 ?

iaguis replied:
Yup, I'll get host's slice name from libsystemd too

jonboulle commented on stage1/init/init.go:
```diff
@@ -295,6 +307,165 @@ func forwardedPorts(pod *Pod) ([]networking.ForwardedPort, error) {
+		// 4c. Create and bind-mount app cgroup directories over themselves so
+		// they stay read-write
+		for _, a := range appHashes {
+			serviceName := types.ShortHash(a.String()) + ".service"
```
there is a helper for this in path.go

jonboulle commented on stage1/init/init.go:
```diff
@@ -295,6 +307,165 @@ func forwardedPorts(pod *Pod) ([]networking.ForwardedPort, error) {
+			serviceName := types.ShortHash(a.String()) + ".service"
+			appCgroup := filepath.Join(subcgroup, serviceName)
+
+			// 4c. Create appCgroup directory
```
the numbering is all out of whack ;-)

jonboulle commented on stage1/init/init.go:
```diff
@@ -369,6 +540,14 @@ func stage1() int {
 
+	appHashes := p.GetAppHashes()
+	s1Root := common.Stage1RootfsPath(p.Root)
+	machineID := "rkt-" + p.UUID.String()
```
please split this out to a helper for pod.go to use too

jonboulle commented on stage1/init/init.go:
```diff
@@ -396,6 +575,38 @@ func stage1() int {
+
+	ret := C.my_sd_pid_get_unit(sd_pid_get_unit, 0, &u)
+	if ret < 0 {
+		err = fmt.Errorf("error calling sd_pid_get_owner_uid: %v", syscall.Errno(-ret))
```
copypasta

iaguis commented:
Updated

alban commented on stage1/init/init.go:
```diff
@@ -295,6 +316,167 @@ func forwardedPorts(pod *Pod) ([]networking.ForwardedPort, error) {
+			if err != nil {
+				return fmt.Errorf("error getting unit name: %v", err)
+			}
+			subcgroup = filepath.Join(cPath, slice, unit, "system.slice")
```
@jonboulle previously commented about `system.slice` but the comment is hidden by github. So let me copy the comment:

> > isn't this counter to https://github.com/coreos/rkt/issues/911#issuecomment-103853978 ?

> Yup, I'll get host's slice name from libsystemd too

iaguis replied:
I get the host slice name from libsystemd but the second "system.slice" is the one in stage1 and we know that's where the apps will be started so I think using it is fine here.

alban replied:
You're right.

alban commented on stage1/init/init.go:
```diff
@@ -369,6 +551,14 @@ func stage1() int {
+	appHashes := p.GetAppHashes()
+	s1Root := common.Stage1RootfsPath(p.Root)
+	machineID := p.GetMachineID()
+	if err := createCgroups(s1Root, machineID, appHashes); err != nil {
```
Does it require systemd-v220 in stage1? How does it behave when rkt is compiled with `RKT_STAGE1_USR_FROM=coreos` (systemd-v215)? If it fails, maybe the app-isolators feature could be disabled when using `flavor=coreos`?

At the moment, `RKT_STAGE1_USR_FROM=src` still build systemd-v219 and I don't know if we will switch to v220 for the next rkt release. Maybe your two commits could be added in stage1/rootfs/usr_from_src/patches/v219:

http://cgit.freedesktop.org/systemd/systemd/commit/?id=54b4755f15438c86991d5a4eaadc47150f7e5a84
http://cgit.freedesktop.org/systemd/systemd/commit/?id=875e1014dd9d55cd0692dcce843598cffb2d09b0

jonboulle replied:
IMHO disabling based on flavo(u)r is reasonable for now.

iaguis replied:
Actually, they work on v219 because systemd-nspawn leaves the cgroups mounted RW :+1:

jonboulle replied:
What about v215?

iaguis replied:
I ninja-edited too late, I meant v215 here. For src-v219 I'll add the patches necessary

alban commented on stage1/init/init.go:
```diff
@@ -396,6 +586,70 @@ func stage1() int {
+}
+
+func getSlice() (slice string, err error) {
+	handle := C.dlopen(C.CString("libsystemd.so"), C.RTLD_LAZY)
```
getSlice() is expensive because it calls dlopen() and dlclose() on each call. getSlice() is called in a loop by createCgroups(). But the returned value is not going to change.

Can you cache the value, either here or in createCgroups()?

alban commented on stage1/init/init.go:
```diff
@@ -295,6 +316,167 @@ func forwardedPorts(pod *Pod) ([]networking.ForwardedPort, error) {
+
+	sc := bufio.NewScanner(f)
+
+	sc.Scan()
```
Add a comment to say that `sc.Scan()` is to skip the first line because it is a comment?
```
#subsys_name	hierarchy	num_cgroups	enabled
```

alban commented:
You could add a couple of functional tests, ideally to check that the isolators are enforced by the kernel. If you want to do that in a later PR, you could at least add a couple of isolators in `tests/image/manifest` with large values and check that the ACIs with the isolators can run without regression. I would like to see if Semaphore can run this on the different flavors of systemd.

alban commented:
To improve the security, individual cgroup knobs could be bind-mounted RW (`cgroup.procs`, `cpu.shares`...) instead of cgroup directories.

This is not blocking for this PR because the pod has CAP_SYS_ADMIN anyway (#576) so malicious apps could already remount the cgroup RW.

jonboulle commented on stage1/init/init.go:
```diff
@@ -295,6 +316,167 @@ func forwardedPorts(pod *Pod) ([]networking.ForwardedPort, error) {
+		}
+
+		// 3d. Re-mount controller read-only to prevent the container modifying host controllers
+		if err := syscall.Mount(cPath, cPath, "", syscall.MS_BIND|syscall.MS_REMOUNT|syscall.MS_NOSUID|syscall.MS_NOEXEC|syscall.MS_NODEV|syscall.MS_RDONLY, ""); err != nil {
```
general comment, can you split lines like this out (e.g. the flags) - otherwise it's way too easy to miss something in a review :-)

jonboulle commented on stage1/init/init.go:
```diff
@@ -295,6 +316,167 @@ func forwardedPorts(pod *Pod) ([]networking.ForwardedPort, error) {
+
+	// 6. Bind-mount cgroup filesystem read-only
+	if err := syscall.Mount(cgroupTmpfs, cgroupTmpfs, "", syscall.MS_BIND|syscall.MS_REMOUNT|syscall.MS_NOSUID|syscall.MS_NOEXEC|syscall.MS_NODEV|syscall.MS_RDONLY, ""); err != nil {
+		return fmt.Errorf("error mounting remounting RO %q: %v", cgroupTmpfs, err)
```
error mounting remounting?

jonboulle commented on stage1/init/init.go:
```diff
@@ -396,6 +586,70 @@ func stage1() int {
 }
 
+func getUnitFileName() (unit string, err error) {
+	handle := C.dlopen(C.CString("libsystemd.so"), C.RTLD_LAZY)
```
This CString needs to be freed afaict.

jonboulle commented on stage1/init/init.go:
```diff
@@ -295,6 +316,167 @@ func forwardedPorts(pod *Pod) ([]networking.ForwardedPort, error) {
+	return symlinks
+}
+
+func createCgroups(root string, machineID string, appHashes []types.Hash) error {
```
A docstring describing the general purpose and need for this would be useful

iaguis commented:
Updated.

I think I addressed all the comments, I'm aware of the code repetition when calling libsystemd functions but hopefully we'll switch to go-systemd soon.

jonboulle commented on stage1/init/init.go:
```diff
@@ -295,6 +316,200 @@ func forwardedPorts(pod *Pod) ([]networking.ForwardedPort, error) {
+			syscall.MS_NODEV |
+			syscall.MS_RDONLY
+		if err := syscall.Mount(cPath, cPath, "", flags, ""); err != nil {
+
```
empty line snuck in

jonboulle commented on stage1/init/init.go:
```diff
@@ -295,6 +316,200 @@ func forwardedPorts(pod *Pod) ([]networking.ForwardedPort, error) {
+		syscall.MS_NODEV |
+		syscall.MS_RDONLY
+	if err := syscall.Mount(cgroupTmpfs, cgroupTmpfs, "", flags, ""); err != nil {
+		return fmt.Errorf("error mounting remounting RO %q: %v", cgroupTmpfs, err)
```
missed this message

jonboulle commented on stage1/init/init.go:
```diff
@@ -369,6 +584,18 @@ func stage1() int {
 		return 3
 	}
 
+	// The systemd version shipped with CoreOS (v215) doesn't allow the
```
can you add a big TODO here please? (at some point I want to modify the test script to actually audit TODOs in the codebase and mandate that they have end dates)

alban replied:
This has been fixed. Thanks :)

alban commented on common/cgroup_util.go:
```diff
@@ -0,0 +1,190 @@
+package common
```
In the commit message, duplicated word: "several several functions"

alban replied:
This has been fixed. Thanks :)

iaguis commented:
Updated

iaguis commented:
systemd v215 prints some nasty error messages when it tries to create cgroups and the fs is RO:

```
Failed to set cpu.shares on /machine.slice/machine-rkt\x2d0db305d3\x2da464\x2d491b\x2d8509\x2d825674e971fd.scope: Read-only file system
Failed to set cpu.cfs_period_us on /machine.slice/machine-rkt\x2d0db305d3\x2da464\x2d491b\x2d8509\x2d825674e971fd.scope: Read-only file system
Failed to set cpu.cfs_quota_us on /machine.slice/machine-rkt\x2d0db305d3\x2da464\x2d491b\x2d8509\x2d825674e971fd.scope: Read-only file system
Failed to set memory.limit_in_bytes on /machine.slice/machine-rkt\x2d0db305d3\x2da464\x2d491b\x2d8509\x2d825674e971fd.scope: Read-only file system
```

This seems to make [TestVolumes fail sometimes](https://semaphoreci.com/coreos/rkt/branches/pull-request-920/builds/15).

These log messages don't appear in flavor=coreos (different default loglevel?) and hence flavor=coreos builds don't fail.

After v219 [systemd prints these errors in the debug loglevel](http://cgit.freedesktop.org/systemd/systemd/commit?id=714e2e1d56b97dcf2ebae2d0447b48f21e38a600)

I tried removing the `--debug` parameter from TestVolumes and [it doesn't fail](https://semaphoreci.com/coreos/rkt/branches/pull-request-920/builds/16).

I think at this point we should wait for #966 to get merged and ignore flavor=src v215

jonboulle commented on common/cgroup_util.go:
```diff
@@ -0,0 +1,179 @@
+func cgEscape(p string) string {
+	needPrefix := false
+
+	if strings.HasPrefix(p, "_") ||
```
switch statement?

jonboulle commented on common/cgroup_util.go:
```diff
@@ -0,0 +1,179 @@
+	validChars = regexp.MustCompile(`[a-zA-Z0-9:\-_\.\\]+`)
+)
+
+// cgEscape implements very minimal escaping for names to be used as file names
```
tests would be great

iaguis replied:
Added tests for cgEscape and SliceToPath

jonboulle replied:
thanks!

jonboulle commented on common/cgroup_util.go:
```diff
@@ -0,0 +1,179 @@
+
+package common
+
+// adapted from systemd/src/shared/cgroup-util.c
```
Maybe a TODO that this can/should move to go-systemd?

jonboulle commented on stage1/init/init.go:
```diff
@@ -297,6 +318,204 @@ func forwardedPorts(pod *Pod) ([]networking.ForwardedPort, error) {
+	for ln, tgt := range symlinks {
+		lnPath := filepath.Join(cgroupTmpfs, ln)
+		if err := os.Symlink(tgt, lnPath); err != nil {
+			return fmt.Errorf("Error creating symlink: %v", err)
```
s/Error/error/

jonboulle commented:
@iaguis so if I understand correctly, you're going to bump this to 220?

alban commented:
@iaguis : #966 has been merged. Can you rebase?

iaguis commented:
Updated.

To clarify a bit:

This feature works on the coreos image because we skip creating the cgroups in rkt and systemd-nspawn v215 doesn't remount them read-only so everything is fine.

From v219 on it should work fine and since we're not testing v215 from src now (#966), tests should pass.

alban commented:
I did some manual tests with user namespaces and cgroups with:
```
sudo systemd-nspawn -D . --private-users=1000:65536
```

Everything in /sys belongs to nobody:nogroup, so it might be a problem for networking applications that needs /sys (`/sys/class/net/`). But that problem is not related to this PR.

But `/sys/fs/cgroup` behaves correctly:
- services outside of the container (e.g. `/sys/fs/cgroup/systemd/system.slice/cups.service`) belongs to nobody so even if `rkt` mistakenly gave read-write access, the container would not be able to mess with it.
- services in the container (e.g. `/sys/fs/cgroup/systemd/machine.slice/machine-debian\x2dtree.scope/`) belong to root from the point of view of the container, or to the mapped user from outside, so the container will be able to read-write them.

So it's a good news :+1: 

/cc @tixxdz

jonboulle commented on common/cgroup_util.go:
```diff
@@ -0,0 +1,185 @@
+		fallthrough
+	case p == "tasks":
+		needPrefix = true
+	default:
```
```
case strings.Contains():
  ...
default:
}

jonboulle commented:
Looks good overall. Unit tests are failing.

iaguis commented:
> Unit tests are failing.

Yeah, they fail on Travis because cgEscape escapes some strings depending on the values of `/sys/fs/cgroup/*` and on the test case we assume `/sys/fs/cgroup/memory` exists. I guess they don't mount any cgroups.

I'll just remove that test case, if `/sys/fs/cgroup/memory` doesn't exist not escaping it is the right thing to do.

alban commented:
LGTM!

I added a note in #845.

Can you open follow-up issues for tests (including checking that the filesystems are mounted with the correct RO flags) and for bind-mounting individual cgroup knobs?

iaguis commented:
Follow-up issues: #981, #982, #983

iaguis commented:
Tests are green now so let's merge this :smile:
